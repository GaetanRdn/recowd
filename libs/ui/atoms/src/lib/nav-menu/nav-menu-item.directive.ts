import { Directive, EventEmitter, HostListener, Output } from '@angular/core';

@Directive({
  selector: '[rcNavMenuItem], rc-nav-menu-item',
  host: { class: 'rc-nav-menu-item', '[class.rc-selected]': 'selected' },
})
export class NavMenuItemDirective {
  public selected = false;

  @Output()
  public readonly selectionChange: EventEmitter<NavMenuItemDirective> = new EventEmitter<NavMenuItemDirective>();

  @HostListener('click')
  private click(): void {
    this.selected = true;
    this.selectionChange.emit(this);
  }
}
