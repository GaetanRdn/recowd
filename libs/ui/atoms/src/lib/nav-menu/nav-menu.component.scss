@mixin rc-nav-menu($config) {
  $typography: map-get($config, typography);
  $colors: map-get($config, colors);

  rc-nav-menu {
    display: flex;
    flex-flow: column;
    min-width: min-content;
    width: 25rem;
    height: 100%;
    position: absolute;
    background-color: map-get(map-get($colors, primary), background-color);
    overflow: hidden;
    left: 0;
  }

  .rc-nav-menu-item {
    @each $key in font-size, font-family, font-weight, line-height {
      #{$key}: map-get(map-get($typography, nav), $key);
    }

    padding: 1.2rem 2rem;
    color: map-get(map-get($colors, primary), color);
    cursor: pointer;
    transition: all 200ms ease-in-out;

    &.rc-selected {
      font-weight: bold;
      border-left: 2px solid map-get(map-get($colors, primary), border-color);
    }
  }
}
