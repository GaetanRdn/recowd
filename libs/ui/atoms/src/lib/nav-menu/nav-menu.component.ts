import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ContentChildren,
  NgZone,
  OnInit,
  Output,
  QueryList,
} from '@angular/core';
import {
  defer,
  filter,
  map,
  merge,
  Observable,
  startWith,
  switchMap,
  take,
  tap,
} from 'rxjs';
import { NavMenuItemDirective } from './nav-menu-item.directive';

@Component({
  selector: 'rc-nav-menu',
  exportAs: 'rcMenu',
  template: ` <ng-content></ng-content> `,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NavMenuComponent implements OnInit {
  @ContentChildren(NavMenuItemDirective)
  public readonly menuItems?: QueryList<NavMenuItemDirective>;

  @Output()
  public readonly selectionChange: Observable<NavMenuItemDirective> = defer(
    () => {
      const items = this.menuItems;

      if (items) {
        return items.changes.pipe(
          startWith(items),
          switchMap(() =>
            merge(
              ...items.map((item: NavMenuItemDirective) => item.selectionChange)
            )
          )
        );
      }

      return this._ngZone.onStable.pipe(
        take(1),
        switchMap(() => this.selectionChange)
      );
    }
  ) as Observable<NavMenuItemDirective>;

  constructor(
    private readonly _ngZone: NgZone,
    private readonly _cdr: ChangeDetectorRef
  ) {}

  public ngOnInit(): void {
    this.selectionChange
      .pipe(
        map((selectedItem: NavMenuItemDirective) =>
          (this.menuItems ?? []).filter(
            (item: NavMenuItemDirective) =>
              item !== selectedItem && item.selected
          )
        ),
        filter((items: NavMenuItemDirective[]) => !!items.length),
        tap(
          (oldSelected: NavMenuItemDirective[]) =>
            (oldSelected[0].selected = false)
        )
      )
      .subscribe();
  }
}
