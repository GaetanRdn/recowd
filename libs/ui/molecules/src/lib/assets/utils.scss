@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
