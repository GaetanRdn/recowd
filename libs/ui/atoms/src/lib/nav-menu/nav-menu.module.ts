import { NgModule } from '@angular/core';
import { NavMenuComponent } from './nav-menu.component';
import { NavMenuItemDirective } from './nav-menu-item.directive';
import { CommonModule } from '@angular/common';

@NgModule({
  exports: [NavMenuComponent, NavMenuItemDirective],
  declarations: [NavMenuComponent, NavMenuItemDirective],
  imports: [CommonModule],
})
export class NavMenuModule {}
