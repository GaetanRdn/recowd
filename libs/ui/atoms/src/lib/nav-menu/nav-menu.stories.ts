import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { NavMenuComponent } from './nav-menu.component';
import { NavMenuItemDirective } from './nav-menu-item.directive';
import { NavMenuModule } from './nav-menu.module';
import { BADGE } from '@geometricpanda/storybook-addon-badges';

export default {
  title: 'atoms/nav-menu',
  component: NavMenuComponent,
  subcomponents: {
    NavMenuItemDirective,
  },
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'nav-menu.component',
  },
  decorators: [
    moduleMetadata({
      imports: [NavMenuModule],
    }),
  ],
} as Meta<NavMenuComponent>;

const template: Story<NavMenuComponent> = () => ({
  template: `<rc-nav-menu>
<rc-nav-menu-item>Item 1</rc-nav-menu-item>
<rc-nav-menu-item>Item 2</rc-nav-menu-item>
<rc-nav-menu-item>Item 3</rc-nav-menu-item>
<rc-nav-menu-item>Item 4</rc-nav-menu-item>
</rc-nav-menu>`,
});

export const Default = template.bind({});
