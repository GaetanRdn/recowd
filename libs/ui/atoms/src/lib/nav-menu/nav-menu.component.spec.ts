import { Component } from '@angular/core';
import { TestBed, waitForAsync } from '@angular/core/testing';
import { NavMenuModule } from './nav-menu.module';
import { TemplateLookup } from '@recowd/test/utils';

describe('NavMenuComponent', () => {
  beforeEach(waitForAsync(() =>
    TestBed.configureTestingModule({
      declarations: [HostComponent],
      imports: [NavMenuModule],
    }).compileComponents()));

  it('should create', () => {
    const templateLookup: TemplateLookup<HostComponent> = new TemplateLookup(
      HostComponent
    );
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should emit event on click', () => {
    // GIVEN
    const templateLookup: TemplateLookup<HostComponent> = new TemplateLookup(
      HostComponent
    );
    templateLookup.detectChanges();
    jest.spyOn(templateLookup.hostComponent, 'selectionChange');

    // WHEN
    templateLookup.query('rc-nav-menu-item').click();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.hostComponent.selectionChange).toBeCalledTimes(1);
  });

  it('should set selected the last clicked item', () => {
    // GIVEN
    const templateLookup: TemplateLookup<HostComponent> = new TemplateLookup(
      HostComponent
    );
    templateLookup.detectChanges();
    jest.spyOn(templateLookup.hostComponent, 'selectionChange');
    templateLookup.query('rc-nav-menu-item').click();
    templateLookup.detectChanges();

    // WHEN
    templateLookup.queryAll('rc-nav-menu-item')[1].click();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.hostComponent.selectionChange).toBeCalledTimes(2);
  });
});

@Component({
  template: ` <rc-nav-menu (selectionChange)="selectionChange($event)">
    <rc-nav-menu-item>Item 1</rc-nav-menu-item>
    <rc-nav-menu-item>Item 2</rc-nav-menu-item>
  </rc-nav-menu>`,
})
class HostComponent {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public selectionChange(_: unknown): void {
    // empty for testing
  }
}
