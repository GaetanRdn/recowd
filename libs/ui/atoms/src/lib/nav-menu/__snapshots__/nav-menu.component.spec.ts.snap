// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenuComponent should create 1`] = `
<rc-nav-menu>
  <rc-nav-menu-item
    class="rc-nav-menu-item"
  >
    Item 1
  </rc-nav-menu-item>
  <rc-nav-menu-item
    class="rc-nav-menu-item"
  >
    Item 2
  </rc-nav-menu-item>
</rc-nav-menu>
`;

exports[`NavMenuComponent should emit event on click 1`] = `
<rc-nav-menu>
  <rc-nav-menu-item
    class="rc-nav-menu-item rc-selected"
  >
    Item 1
  </rc-nav-menu-item>
  <rc-nav-menu-item
    class="rc-nav-menu-item"
  >
    Item 2
  </rc-nav-menu-item>
</rc-nav-menu>
`;

exports[`NavMenuComponent should set selected the last clicked item 1`] = `
<rc-nav-menu>
  <rc-nav-menu-item
    class="rc-nav-menu-item"
  >
    Item 1
  </rc-nav-menu-item>
  <rc-nav-menu-item
    class="rc-nav-menu-item rc-selected"
  >
    Item 2
  </rc-nav-menu-item>
</rc-nav-menu>
`;
